### PR TITLE
PIL-492 - Create confirm client details page

### PR DESCRIPTION
**Pull request by TevynAllen**

### CODE REVIEW

File: app/controllers/AgentController.scala [hunk 1]
```diff
@@ -16,67 +16,77 @@
 
 package controllers
 
+import cats.implicits._
 import config.FrontendAppConfig
 import connectors.UserAnswersConnectors
-import controllers.actions.{DataRequiredAction, DataRetrievalAction, FeatureFlagActionFactory, IdentifierAction}
+import controllers.AgentController.VerifyAgentClientPredicate
+import controllers.actions.{AgentIdentifierAction, DataRequiredAction, DataRetrievalAction, FeatureFlagActionFactory, IdentifierAction}
```
Comment (jassalrichy) [inline app/controllers/AgentController.scala:23]: remove unused import IdentifierAction

File: app/controllers/AgentController.scala [hunk 2]
```diff
@@ -16,67 +16,77 @@
...
 import models.InternalIssueError
-import pages.AgentClientPillar2ReferencePage
+import pages.{AgentClientOrganisationNamePage, AgentClientPillar2ReferencePage}
 import play.api.i18n.I18nSupport
 import play.api.libs.json.Json
 import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
 import services.SubscriptionService
+import uk.gov.hmrc.auth.core.AuthProvider.GovernmentGateway
+import uk.gov.hmrc.auth.core.authorise.Predicate
+import uk.gov.hmrc.auth.core.{AuthConnector, AuthProviders, AuthorisedFunctions, Enrolment}
```
Comment (jassalrichy) [inline app/controllers/AgentController.scala:33]: remove unused imports.

File: app/controllers/actions/AgentIdentifierAction.scala
Comment (MarjanMubarok) [inline app/controllers/actions/AgentIdentifierAction.scala:1]: Can this be unit tested? Including the placeholders for PIL 505 and 922?

File: app/controllers/actions/AgentIdentifierAction.scala [hunk 3]
```diff
@@ -18,52 +18,69 @@ package controllers.actions
...
+
+        authorised(agentPredicate)
+          .retrieve(Retrievals.internalId and Retrievals.allEnrolments and Retrievals.affinityGroup and Retrievals.credentialRole) {
 
-    authorised(AuthProviders(GovernmentGateway) and ConfidenceLevel.L50)
-      .retrieve(Retrievals.internalId and Retrievals.allEnrolments and Retrievals.affinityGroup and Retrievals.credentialRole) {
+            case Some(internalId) ~ enrolments ~ Some(Agent) ~ _ =>
+              Future.successful(Right(IdentifierRequest(request, internalId, enrolments = enrolments.enrolments, isAgent = true)))
+            case _ ~ _ ~ Some(Organisation) ~ Some(User) =>
+              Future.successful(Left(Redirect(routes.UnauthorisedController.onPageLoad)))
```
Comment (jassalrichy) [inline app/controllers/actions/AgentIdentifierAction.scala:58]: Check redirect to UnauthorisedController. Screen has missing content.

File: app/controllers/actions/AgentIdentifierAction.scala [hunk 4]
```diff
@@ -18,52 +18,69 @@ package controllers.actions
...
+              logger.warn(s"[Session ID: ${Pillar2SessionKeys.sessionId(hc)}] - Unable to retrieve internal id or affinity group")
+              Future.successful(Left(Redirect(routes.UnauthorisedController.onPageLoad)))
+          } recover {
+          case _: NoActiveSession =>
+            Left(Redirect(config.loginUrl, Map("continue" -> Seq(config.loginContinueUrl))))
+          case e: InsufficientEnrolments if e.reason == HMRC_PILLAR2_ORG_KEY =>
+            Left(Redirect(routes.UnderConstructionController.onPageLoad)) // PIL-505
+          case _: InternalError => Left(Redirect(routes.UnderConstructionController.onPageLoad)) // PIL-922
+          case _: AuthorisationException =>
+            Left(Redirect(routes.UnauthorisedController.onPageLoad))
```
Comment (jassalrichy) [inline app/controllers/actions/AgentIdentifierAction.scala:70]: See comment above.

File: app/controllers/actions/AgentIdentifierAction.scala [hunk 5]
```diff
@@ -18,52 +18,69 @@ package controllers.actions
...
+        implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+
+        authorised(agentPredicate)
+          .retrieve(Retrievals.internalId and Retrievals.allEnrolments and Retrievals.affinityGroup and Retrievals.credentialRole) {
 
-    authorised(AuthProviders(GovernmentGateway) and ConfidenceLevel.L50)
-      .retrieve(Retrievals.internalId and Retrievals.allEnrolments and Retrievals.affinityGroup and Retrievals.credentialRole) {
+            case Some(internalId) ~ enrolments ~ Some(Agent) ~ _ =>
+              Future.successful(Right(IdentifierRequest(request, internalId, enrolments = enrolments.enrolments, isAgent = true)))
+            case _ ~ _ ~ Some(Organisation) ~ Some(User) =>
```
[the remark this answers is not in the document]
Comment (TevynAllen) [reply app/controllers/actions/AgentIdentifierAction.scala]: Theres a test for Organisation which is enough

File: app/controllers/actions/AgentIdentifierAction.scala (hunk 3 above)
Comment (TevynAllen) [reply app/controllers/actions/AgentIdentifierAction.scala]: This page isn't UCD designed and generated by scaffold 
we have a UCD ticket to define error screens for it so for now we can use placeholder error pages - https://jira.tools.tax.service.gov.uk/browse/PIL-955

File: app/controllers/actions/AgentIdentifierAction.scala (hunk 4 above)
Comment (TevynAllen) [reply app/controllers/actions/AgentIdentifierAction.scala:70]: ditto

File: app/controllers/actions/AgentIdentifierAction.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/controllers/actions/AgentIdentifierAction.scala]: Screenshot from 2024-04-29 10-06-50

But, we are not testing the case as shown in this screen shot.

**Review by jassalrichy — DISMISSED** [review]

**Review by jassalrichy — APPROVED** [review]

**Pull request merged**